### PR TITLE
Auth filters: allow specify mTLS mode not through endpoint port

### PR DESCRIPTION
**Pull request by lambdai**
#24540

The goal of Better transport security is to create a tunnel inbound listener and *enforce mTLS for all inbound traffic there*.

Ideally we eventually push all traffic to that listener but at this moment we consider only HTTP traffic.
 
This PR allows to create an authn filter not attached to any endpoint port.

Follow up: utilize the nil ServiceInstance when building the filter chain for BTS tunnel listener.
Signed-off-by: Yuchen Dai <silentdai@gmail.com>

[x] Security

### CODE REVIEW

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 1]
```diff
@@ -37,9 +38,13 @@ func NewPlugin() plugin.Plugin {
 
 // OnInboundFilterChains setups filter chains based on the authentication policy.
 func (Plugin) OnInboundFilterChains(in *plugin.InputParams) []networking.FilterChain {
+	var tlsGetter authn.TlsModeGetter = authn.StrictTlsModeValue{}
+	if in.ServiceInstance != nil {
```
[the remark this answers is not in the document]
Comment (lambdai) [reply pilot/pkg/networking/plugin/authn/authentication.go:42]: @costinm Thanks for the feedback!

Yes, this is only for BTS where only STRICT make sense.

The previous in.ServiceInstance must be not nil. Before this PR the endpoint port in ServiceInstance is always used. So no existing case is broken.

I believe your concern is more about future regression for accidental use other than BTS. I can add a runtime validation check on port 15008 as you suggest.

**Comment by incfly** [comment]
I am bit confused by this PR.

This seems only start some initial structure change and we should expect some follow-up to complete BTS? Can you explain how this PR will help and evolve into for the end state by then?

Some sample config snippet might help. If that's already covered somewhere, just put the link I'll read more.

**Comment by lambdai** [comment]
@incfly Thanks for the comment!

The entire BTS design doc is addressed in #24540

In current phase we consider only a listener with strict mTLS and is prepared to terminate all HTTP traffic into the sidecar.

1. Now that BTS requires traffic hitting the tunnel listener is mTLS, we could either specialize an authn plugin on BTS listener reusing the existing
authn plugin as much as possible.

2. As per discussion with @yangminzhu and you, the auth filters are uniformed if excluding mTLS regardless auth filter is generated for service port 80 or service port 8080.
The only usage of the port is to figure out the mTLS enum value. 

The approach is to pass either the endpoint, or the mTLS requirement by BTS(mTLS strict).

The interface is merely the implementation.

In the follow BTS PR, this OnInboundFilterChains will be used to manufacture a reasonable filters.

**Comment by lambdai** [comment]
@incfly can I land it first and do validation check in follow up PRs?

**Comment by incfly** [comment]
So I understand correctly, this PR is to reuse policy applier related code, create a STRICT mode inbound listener, which is used for tunneling? At this phase, httpbin:8000 if configured PERMISSIVE, we still have two filter chain on port 15006 respectively?

If that's the only purpose, can we fake a internal peer authentication policy configuring 16000 port as STRICT? This way all the security code can remain the same.

Reason I propose that is so that during the rollout of BTS (across at lease releases I assume), we have much higher confidence the existing policy resolution logic remains the same. BTS and current PeerAuthn  can stay decoupled.

**Comment by lambdai** [comment]
Sorry I don't catch up your concern.
The complexity originates from the system: Both tunnel listener and other listeners depends on the same plugin, and it's not only mTLS plugin. 
I can replicated the code, but both of us will experience the pain that non of us will be notified when auth policy changes. 
The OnInboundPassthroughFilterChains is an example to support my proposal (Let's investigate later)

You are absolutely right regarding `NewPolicyApplier`. I should have erased the `ServiceInstance` earlier. See the latest commit.

Now we are crystal clear the `ServiceInstance` is not required in PolicyApplier. 
Unfortunately the plugin.InputParam has to contain it for other plugin functionality.

**Pull request closed without merging**